lib/Transforms/IPO/StructAnalysisTransformAnalyzer.cpp

```diff
@@ -0,0 +1,873 @@
+  CloseProximityRelations.resize(NumElements);
+  double MaxWCP = 0;
+  FieldNumType MaxI, MaxJ;
+  for (unsigned i = 0; i < NumElements; i++) {
+    CloseProximityRelations[i].resize(NumElements);
+    for (unsigned j = 0; j < NumElements; j++) {
+      CloseProximityRelations[i][j] = calculateCloseProximity(i, j);
+      if (i < j && CloseProximityRelations[i][j] > MaxWCP) {
+        MaxWCP = CloseProximityRelations[i][j];
+        MaxI = i;
+        MaxJ = j;
+      }
+    }
+  }
+  if (MaxWCP < 1e-3) {
+    Eligibility = false;
+    return;
+  }
+  DEBUG(dbgs() << "CPG for field reordering recommendation\n");
+  for (unsigned i = 0; i < NumElements; i++) {
+    DEBUG(dbgs() << "F" << i + 1 << "\t");
+    for (unsigned j = 0; j < NumElements; j++)
+      DEBUG(dbgs() << DEBUG_PRINT_COUNT(CloseProximityRelations[i][j]) << "\t");
+    DEBUG(dbgs() << "\n");
+  }
+
+  // FIXME: Need to reconsider if the order of the pair makes differences
+  NewOrder.push_back(MaxI);
+  NewOrder.push_back(MaxJ);
+  FieldsToTransform.erase(MaxI);
+  FieldsToTransform.erase(MaxJ);
+  DEBUG_WITH_TYPE(DEBUG_TYPE_REORDER, dbgs() << "Use F" << MaxI + 1 << " and F"
+                                             << MaxJ + 1
+                                             << " as seeds in reordering\n");
+  assert(MaxWCP == getWCP());
+  DEBUG_WITH_TYPE(DEBUG_TYPE_REORDER,
+                  dbgs() << "Current WCP: " << DEBUG_PRINT_COUNT(MaxWCP)
+                         << "\n");
+}
+
+// Calculate Close Proximity in this reordering algorithm. This algorithm
+// takes in a pair of CP relation of (count, distance) and return a single
+// floating point number that is count/distance. Special cases is that when
+// two fields are the same, or they can't fit into the same cache block,
+// return 0. When distance is smaller than 1 byte, return the count as if
+// the distance is 1 byte.
+double
+FieldReorderAnalyzer::calculateCloseProximity(FieldNumType Field1,
+                                              FieldNumType Field2) const {
+  if (Field1 == Field2)
+    return 0;
+  if (!canFitInOneCacheBlock(Field1, Field2))
+    return 0;
+  auto *Pair = CPBuilder->getCloseProximityPair(Field1, Field2);
+  if (Pair->first < 1e-3)
+    return 0;
+  if (Pair->second < 1)
+    return Pair->first;
+  return Pair->first / Pair->second;
+}
+
+// Function to check if any part of the two fields can fit on the same
+// cache block or not. When two fields are too big to fit on one cache block,
+// we check if any part of the first field can be on the same cache block
+// of the second field.
+// This function is used to measure if it could be beneficial to put two
+// fields together because bringing one field will load another field
+// into the cache if the function returns true
+bool FieldReorderAnalyzer::canFitInOneCacheBlock(FieldNumType Field1,
+                                                 FieldNumType Field2) const {
+  if (FieldSizes[Field1] % CacheBlockSize == 0)
```

Comment:
No, it's correct but might need some changes. When defining this function, I thought need both of the fields to check if their size can fit in one cache block. But later when I found that one field can be larger than a cache block itself, we move on to check if any part of the two fields can fit within one cache block. In this case, only size of Field1 seems to matter because as long as it doesn't take all the cache block, Field2 can fit on the same cache block.